File: solutions/javascript/robot-simulator/1/robot-simulator.js
export class InvalidInputError extends Error {
  constructor(message) {
    super();
    this.message = message || 'Invalid Input';
  }
}

export class Robot {
  #bearing;
  #x;
  #y;

  constructor(options) {
    this.#bearing = "north";
    this.#x = 0;
    this.#y = 0;
  }
  
  get bearing() {
    return this.#bearing.slice(0);
  }

  get coordinates() {
    return [this.#x, this.#y];
  }

  place({ direction, x, y }) {
    const validDirections = ["north", "east", "south", "west"];
    if (!validDirections.includes(direction))
      throw new InvalidInputError;
    this.#bearing = direction;
    this.#x = x;
    this.#y = y;
  }

  #turnLeft() {
    switch (this.#bearing) {
      case "north":
        this.#bearing = "west";
        return;
      case "east":
        this.#bearing = "north";
        return;
      case "south":
        this.#bearing = "east";
        return;
      case "west":
        this.#bearing = "south";
        return;
    }
  }

  #turnRight() {
    switch (this.#bearing) {
      case "north":
        this.#bearing = "east";
        return;
      case "east":
        this.#bearing = "south";
        return;
      case "south":
        this.#bearing = "west";
        return;
      case "west":
        this.#bearing = "north";
        return;
    }
  }

  #advance() {
    switch (this.#bearing) {
      case "north":
        this.#y++;
        return;
      case "south":
        this.#y--;
        return;
      case "east":
        this.#x++;
        return;
      case "west":
        this.#x--;
        return;
    }
  }

  evaluate(instructions) {
    if (!instructions || typeof instructions[Symbol.iterator] !== "function")
      return;
    
    [...instructions].forEach(instruction => {
      switch (instruction) {
        case "R":
          this.#turnRight();
          break;
        case "L":
          this.#turnLeft();
          break;
        case "A":
          this.#advance();
          break;
        default:
          throw new InvalidInputError;
      }
    });
  }
}

File: solutions/cpp/robot-simulator/1/robot_simulator.cpp
#include "robot_simulator.h"

namespace robot_simulator {
   Robot::Robot() : m_position{}, m_bearing{Bearing::NORTH} {}

   Robot::Robot(std::pair<int, int> position, Bearing bearing) :
        m_position{std::move(position)}, 
        m_bearing{std::move(bearing)} {}

   auto Robot::get_position() const -> std::pair<int, int> {
       return m_position;
   }

   auto Robot::get_bearing() const -> Bearing {
       return m_bearing;
   }

   auto Robot::turn_right() -> void {
       switch (m_bearing) {
           case Bearing::NORTH:
             m_bearing = Bearing::EAST;
             break;
           case Bearing::EAST:
             m_bearing = Bearing::SOUTH;
             break;
           case Bearing::SOUTH:
             m_bearing = Bearing::WEST;
             break;
           case Bearing::WEST:
             m_bearing = Bearing::NORTH;
             break;
       }
   }

   auto Robot::turn_left() -> void {
       switch (m_bearing) {
           case Bearing::NORTH:
             m_bearing = Bearing::WEST;
             break;
           case Bearing::EAST:
             m_bearing = Bearing::NORTH;
             break;
           case Bearing::SOUTH:
             m_bearing = Bearing::EAST;
             break;
           case Bearing::WEST:
             m_bearing = Bearing::SOUTH;
             break;
       }
   }

   auto Robot::advance() -> void {
       switch (m_bearing) {
         case Bearing::NORTH:
           m_position.second++;
           break;
         case Bearing::SOUTH:
           m_position.second--;
           break;
         case Bearing::EAST:
           m_position.first++;
           break;
         case Bearing::WEST:
           m_position.first--;
           break;
       }
   }

   auto Robot::execute_sequence(const std::string& commands) -> void{
       for (const auto& command : commands) {
           switch (command) {
               case 'R':
                  turn_right();
                  break;
               case 'L':
                  turn_left();     
                  break;
               case 'A':
                  advance();     
                  break;
            }
        }
    }
}  // robot_simulator
